ipld-schema.ipldsch

```diff
@@ -0,0 +1,46 @@
+type UCAN struct {
+  version String
+
+  issuer PublicKey
+  audience PublicKey
+  signature Signature
+
+  capabilities [Capability]
+  -- All proofs are links, however you could still inline proof
+  -- by using CID with identity hashing algorithm
+  proofs [&UCAN]
+  expiration Int
+
+  facts [Fact]
+  nonce optional String
+  notBefore optional Int
+} representation map {
+  field facts default []
+  field proofs default []
+}
+
+
+type Capability struct {
+  with Resource
+  can Ability
+  -- can have arbitrary other fields
```

Comment:
@expede given IPLD schema limitation I'm inclined to change this as follows

```suggestion
  caveats { String: Any }
```

Mismatch with JWT / JS representation is unfortunate however and I would much rather amend UCAN spec to just to push other fields into substructure.

Alternatively we could define capability as

```ipldsch
type Capability { String: Any }
```

But personally I find that option even worth.

Thoughts ?

Reply:
Yup I'm fine with the overall shape. I'm less certain about the word `caveats` since the enclosed information may include information beyond what are traditionally considered caveats.

In the current draft spec, we use `ext` to signal "extra" fields, but I'd love to find a more descriptive term. To some degree, this feels a bit like xattrs/xargs, so maybe something as short as `x` is fine?

Reply:
Originally I was going for "constraints" instead but then I learned that [zcaps](https://w3c-ccg.github.io/zcap-spec/#caveats) use `caveats` instead and according to one of [dictionary definitions](https://www.merriam-webster.com/dictionary/caveat) it is

> an explanation to prevent misinterpretation

which seem to be exactly it, or maybe I'm failing to see what kind of information would fail to fit that definition.

For what it's worth I have also considered `if`, `etc`, `and`, `when` as an alternative names, yet I still like "caveats" better. That said I'm any name will do.